### ISSUE TITLE
JWT Support doesn't preserve Roles and Permissions [Testcase included]

### ISSUE THREAD

**Issue by theomega** [pac4j/pac4j#485]
If you generate a JWT for a `DbProfile`, the information about roles and permissions is lost.
See this failing test-case:

```
import org.junit.Test;
import org.pac4j.core.profile.UserProfile;
import org.pac4j.http.credentials.HttpCredentials;
import org.pac4j.http.credentials.TokenCredentials;
import org.pac4j.jwt.credentials.authenticator.JwtAuthenticator;
import org.pac4j.jwt.profile.JwtGenerator;
import org.pac4j.sql.profile.DbProfile;

import static org.assertj.core.api.Assertions.*;


public class JwtTest {
    @Test
    public void test() {
        DbProfile input = new DbProfile();
        input.setId("admin");
        input.addRole("ADMIN_ROLE");
        input.addPermission("SOME_PERMISSION");

        JwtGenerator generator = new JwtGenerator("12345678901234567890123456789012");
        String token = generator.generate(input);
        TokenCredentials credentials = new TokenCredentials(token, "HeaderClient");

        JwtAuthenticator authenticator = new JwtAuthenticator("12345678901234567890123456789012");
        authenticator.validate(credentials);

        assertThat(credentials.getUserProfile().getId()).isEqualTo("admin");
        assertThat(credentials.getUserProfile().getRoles()).contains("ADMIN_ROLE");
        assertThat(credentials.getUserProfile().getPermissions()).contains("SOME_PERMISSION");
    }

}
```

The JWT generator doesn't even store this information in the token, so it also cannot be retrieved again. Is this desired?

**Comment by leleuj** [comment]
You raise a good point: this is a use case which has not been fully thought.

When I added roles / permissions to pac4j, I had in mind that they would be computed from profiles (via an `AuthorizationGenerator`), but through JWT or when the account registration will be available (https://github.com/pac4j/pac4j/issues/339), the roles / permissions must be restored.

OK. So I think the easiest way to do that is to save roles / permissions as specific attributes of the user profile (with specific keys like: "_$internal.roles" and "_$internal.permissions").

I will handle that for version 1.8.6.

**Comment by leleuj** [comment]
For 1.8.x: https://github.com/pac4j/pac4j/pull/487

**Comment by leleuj** [comment]
For master: https://github.com/pac4j/pac4j/pull/488

**Comment by leleuj** [comment]
@theomega : I just released pac4j v1.8.6.
